File: src/services/spotifyAuthService.ts
import * as Crypto from 'expo-crypto';
import * as SecureStore from 'expo-secure-store';
import { SpotifyToken, SpotifyUser } from '../types';

// ---------------------------------------------------------------------------
// Configuration – replace CLIENT_ID with your Spotify Developer Dashboard ID
// ---------------------------------------------------------------------------
export const SPOTIFY_CLIENT_ID = process.env.EXPO_PUBLIC_SPOTIFY_CLIENT_ID ?? '';

export const SPOTIFY_SCOPES = [
  'user-read-email',
  'user-read-private',
  'streaming',
  'user-read-playback-state',
  'user-modify-playback-state',
  'playlist-read-private',
  'playlist-read-collaborative',
].join(' ');

const SPOTIFY_AUTH_ENDPOINT = 'https://accounts.spotify.com/authorize';
const SPOTIFY_TOKEN_ENDPOINT = 'https://accounts.spotify.com/api/token';
const SPOTIFY_USER_ENDPOINT = 'https://api.spotify.com/v1/me';

const SECURE_STORE_KEY = 'spotify_token';

// ---------------------------------------------------------------------------
// PKCE helpers
// ---------------------------------------------------------------------------

/**
 * Generates a cryptographically random code verifier (43-128 chars, URL-safe).
 */
export async function generateCodeVerifier(): Promise<string> {
  const randomBytes = await Crypto.getRandomBytesAsync(32);
  // Base64url-encode without padding
  return Buffer.from(randomBytes)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

/**
 * Derives a code challenge from a code verifier using SHA-256.
 */
export async function generateCodeChallenge(verifier: string): Promise<string> {
  const digest = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    verifier,
    { encoding: Crypto.CryptoEncoding.BASE64 },
  );
  // Convert standard Base64 to Base64url
  return digest
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=/g, '');
}

// ---------------------------------------------------------------------------
// Authorization URL
// ---------------------------------------------------------------------------

/**
 * Builds the Spotify OAuth 2.0 PKCE authorization URL.
 */
export function buildAuthorizationUrl(
  redirectUri: string,
  codeChallenge: string,
  state: string,
): string {
  const params = new URLSearchParams({
    response_type: 'code',
    client_id: SPOTIFY_CLIENT_ID,
    scope: SPOTIFY_SCOPES,
    redirect_uri: redirectUri,
    state,
    code_challenge_method: 'S256',
    code_challenge: codeChallenge,
  });
  return `${SPOTIFY_AUTH_ENDPOINT}?${params.toString()}`;
}

// ---------------------------------------------------------------------------
// Token exchange
// ---------------------------------------------------------------------------

/**
 * Exchanges an authorization code for an access + refresh token pair.
 */
export async function exchangeCodeForToken(
  code: string,
  codeVerifier: string,
  redirectUri: string,
): Promise<SpotifyToken> {
  const body = new URLSearchParams({
    grant_type: 'authorization_code',
    code,
    redirect_uri: redirectUri,
    client_id: SPOTIFY_CLIENT_ID,
    code_verifier: codeVerifier,
  });

  const response = await fetch(SPOTIFY_TOKEN_ENDPOINT, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: body.toString(),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Token exchange failed (${response.status}): ${text}`);
  }

  const data = await response.json() as {
    access_token: string;
    refresh_token: string;
    expires_in: number;
    scope: string;
  };

  return {
    accessToken: data.access_token,
    refreshToken: data.refresh_token,
    expiresAt: Date.now() + data.expires_in * 1000,
    scope: data.scope,
  };
}

// ---------------------------------------------------------------------------
// Token refresh
// ---------------------------------------------------------------------------

/**
 * Uses a refresh token to obtain a new access token.
 */
export async function refreshAccessToken(refreshToken: string): Promise<SpotifyToken> {
  const body = new URLSearchParams({
    grant_type: 'refresh_token',
    refresh_token: refreshToken,
    client_id: SPOTIFY_CLIENT_ID,
  });

  const response = await fetch(SPOTIFY_TOKEN_ENDPOINT, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: body.toString(),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Token refresh failed (${response.status}): ${text}`);
  }

  const data = await response.json() as {
    access_token: string;
    refresh_token?: string;
    expires_in: number;
    scope: string;
  };

  return {
    accessToken: data.access_token,
    // Spotify may or may not return a new refresh token
    refreshToken: data.refresh_token ?? refreshToken,
    expiresAt: Date.now() + data.expires_in * 1000,
    scope: data.scope,
  };
}

// ---------------------------------------------------------------------------
// Secure token storage
// ---------------------------------------------------------------------------

/**
 * Persists the token bundle in device secure storage.
 */
export async function saveToken(token: SpotifyToken): Promise<void> {
  await SecureStore.setItemAsync(SECURE_STORE_KEY, JSON.stringify(token));
}

/**
 * Retrieves the stored token bundle, or null if none exists.
 */
export async function getStoredToken(): Promise<SpotifyToken | null> {
  const raw = await SecureStore.getItemAsync(SECURE_STORE_KEY);
  if (!raw) return null;
  return JSON.parse(raw) as SpotifyToken;
}

/**
 * Removes the stored token from secure storage.
 */
export async function clearToken(): Promise<void> {
  await SecureStore.deleteItemAsync(SECURE_STORE_KEY);
}

// ---------------------------------------------------------------------------
// User profile
// ---------------------------------------------------------------------------

/**
 * Fetches the Spotify profile for the authenticated user.
 * Automatically refreshes the token if it has expired.
 */
export async function getCurrentUser(token: SpotifyToken): Promise<{ user: SpotifyUser; token: SpotifyToken }> {
  let activeToken = token;

  // Refresh if the access token is expired (with a 30-second buffer)
  if (Date.now() >= activeToken.expiresAt - 30_000) {
    activeToken = await refreshAccessToken(activeToken.refreshToken);
    await saveToken(activeToken);
  }

  const response = await fetch(SPOTIFY_USER_ENDPOINT, {
    headers: { Authorization: `Bearer ${activeToken.accessToken}` },
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch user profile (${response.status})`);
  }

  const data = await response.json() as {
    id: string;
    display_name: string;
    email: string;
    images?: Array<{ url: string }>;
  };

  const user: SpotifyUser = {
    id: data.id,
    displayName: data.display_name,
    email: data.email,
    imageUrl: data.images?.[0]?.url,
  };

  return { user, token: activeToken };
}

// ---------------------------------------------------------------------------
// Logout
// ---------------------------------------------------------------------------

/**
 * Clears all locally stored Spotify credentials.
 */
export async function logout(): Promise<void> {
  await clearToken();
}

File: maui/DanceClassApp/Services/SpotifyAuthService.cs
using System.Net.Http.Json;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Serialization;
using DanceClassApp.Models;

namespace DanceClassApp.Services;

/// <summary>
/// Implements <see cref="ISpotifyAuthService"/> using MAUI's
/// <see cref="WebAuthenticator"/> for the browser redirect and
/// <see cref="SecureStorage"/> for token persistence.
///
/// Mirrors <c>spotifyAuthService.ts</c>. Replace <c>ClientId</c> with your
/// Spotify Developer Dashboard application client ID.
/// </summary>
public class SpotifyAuthService : ISpotifyAuthService
{
    // -------------------------------------------------------------------------
    // Configuration – replace with your Spotify Developer Dashboard Client ID.
    // -------------------------------------------------------------------------

    /// <summary>
    /// Spotify OAuth application Client ID.
    /// Set the SPOTIFY_CLIENT_ID environment variable or replace this value directly.
    /// </summary>
    public static string ClientId =>
        Environment.GetEnvironmentVariable("SPOTIFY_CLIENT_ID") ?? string.Empty;

    private static readonly string[] Scopes =
    [
        "user-read-email",
        "user-read-private",
        "streaming",
        "user-read-playback-state",
        "user-modify-playback-state",
        "playlist-read-private",
        "playlist-read-collaborative",
    ];

    private const string AuthEndpoint = "https://accounts.spotify.com/authorize";
    private const string TokenEndpoint = "https://accounts.spotify.com/api/token";
    private const string UserEndpoint = "https://api.spotify.com/v1/me";
    private const string SecureStoreKey = "spotify_token";

    // Redirect URI registered in the Spotify Developer Dashboard.
    // Must match the custom URI scheme declared in AndroidManifest.xml / Info.plist.
    private const string RedirectUri = "danceclassapp://callback";

    private readonly HttpClient _http;

    public SpotifyAuthService(HttpClient http)
    {
        _http = http;
    }

    // -------------------------------------------------------------------------
    // PKCE helpers
    // -------------------------------------------------------------------------

    private static string GenerateCodeVerifier()
    {
        var bytes = RandomNumberGenerator.GetBytes(32);
        return Convert.ToBase64String(bytes)
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=');
    }

    private static string GenerateCodeChallenge(string verifier)
    {
        var bytes = SHA256.HashData(Encoding.ASCII.GetBytes(verifier));
        return Convert.ToBase64String(bytes)
            .Replace('+', '-')
            .Replace('/', '_')
            .TrimEnd('=');
    }

    // -------------------------------------------------------------------------
    // Login
    // -------------------------------------------------------------------------

    /// <inheritdoc />
    public async Task<(SpotifyUser User, SpotifyToken Token)> LoginAsync()
    {
        if (string.IsNullOrWhiteSpace(ClientId))
            throw new InvalidOperationException(
                "Spotify Client ID is not configured. Set the SPOTIFY_CLIENT_ID environment variable.");

        var verifier = GenerateCodeVerifier();
        var challenge = GenerateCodeChallenge(verifier);
        var state = GenerateCodeVerifier(); // random CSRF state

        var authUrl = BuildAuthorizationUrl(challenge, state);
        var callbackUri = new Uri(RedirectUri);

        // Opens the system browser and waits for the callback deep-link.
        var result = await WebAuthenticator.Default.AuthenticateAsync(
            new Uri(authUrl), callbackUri);

        if (!result.Properties.TryGetValue("code", out var code) || string.IsNullOrEmpty(code))
            throw new InvalidOperationException("No authorization code received from Spotify.");

        if (result.Properties.TryGetValue("state", out var returnedState)
            && returnedState != state)
            throw new InvalidOperationException("OAuth state mismatch – possible CSRF attack.");

        if (result.Properties.TryGetValue("error", out var error) && !string.IsNullOrEmpty(error))
            throw new InvalidOperationException($"Spotify auth error: {error}");

        var token = await ExchangeCodeForTokenAsync(code, verifier);
        await SaveTokenAsync(token);

        var (user, freshToken) = await GetCurrentUserAsync(token);
        return (user, freshToken);
    }

    private string BuildAuthorizationUrl(string codeChallenge, string state)
    {
        var scopeStr = string.Join(' ', Scopes);
        var query = new Dictionary<string, string>
        {
            ["response_type"] = "code",
            ["client_id"] = ClientId,
            ["scope"] = scopeStr,
            ["redirect_uri"] = RedirectUri,
            ["state"] = state,
            ["code_challenge_method"] = "S256",
            ["code_challenge"] = codeChallenge,
        };

        var queryString = string.Join("&",
            query.Select(kv =>
                $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));

        return $"{AuthEndpoint}?{queryString}";
    }

    // -------------------------------------------------------------------------
    // Token exchange
    // -------------------------------------------------------------------------

    private async Task<SpotifyToken> ExchangeCodeForTokenAsync(string code, string codeVerifier)
    {
        var body = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["grant_type"] = "authorization_code",
            ["code"] = code,
            ["redirect_uri"] = RedirectUri,
            ["client_id"] = ClientId,
            ["code_verifier"] = codeVerifier,
        });

        var response = await _http.PostAsync(TokenEndpoint, body);
        if (!response.IsSuccessStatusCode)
        {
            var text = await response.Content.ReadAsStringAsync();
            throw new HttpRequestException(
                $"Token exchange failed ({(int)response.StatusCode}): {text}");
        }

        var data = await response.Content.ReadFromJsonAsync<SpotifyTokenResponse>()
                   ?? throw new HttpRequestException("Failed to deserialize token response");

        return ToSpotifyToken(data);
    }

    // -------------------------------------------------------------------------
    // Token refresh
    // -------------------------------------------------------------------------

    /// <inheritdoc />
    public async Task<SpotifyToken> RefreshAccessTokenAsync(string refreshToken)
    {
        var body = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refreshToken,
            ["client_id"] = ClientId,
        });

        var response = await _http.PostAsync(TokenEndpoint, body);
        if (!response.IsSuccessStatusCode)
        {
            var text = await response.Content.ReadAsStringAsync();
            throw new HttpRequestException(
                $"Token refresh failed ({(int)response.StatusCode}): {text}");
        }

        var data = await response.Content.ReadFromJsonAsync<SpotifyTokenResponse>()
                   ?? throw new HttpRequestException("Failed to deserialize token response");

        // Spotify may or may not return a new refresh token
        if (string.IsNullOrEmpty(data.RefreshToken))
            data.RefreshToken = refreshToken;

        return ToSpotifyToken(data);
    }

    private static SpotifyToken ToSpotifyToken(SpotifyTokenResponse data) => new()
    {
        AccessToken = data.AccessToken,
        RefreshToken = data.RefreshToken ?? string.Empty,
        ExpiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + data.ExpiresIn * 1_000L,
        Scope = data.Scope ?? string.Empty,
    };

    // -------------------------------------------------------------------------
    // User profile
    // -------------------------------------------------------------------------

    /// <inheritdoc />
    public async Task<(SpotifyUser User, SpotifyToken Token)> GetCurrentUserAsync(SpotifyToken token)
    {
        var activeToken = token;

        if (activeToken.IsExpired)
        {
            activeToken = await RefreshAccessTokenAsync(activeToken.RefreshToken);
            await SaveTokenAsync(activeToken);
        }

        using var request = new HttpRequestMessage(HttpMethod.Get, UserEndpoint);
        request.Headers.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", activeToken.AccessToken);

        var response = await _http.SendAsync(request);
        if (!response.IsSuccessStatusCode)
            throw new HttpRequestException(
                $"Failed to fetch user profile ({(int)response.StatusCode})");

        var data = await response.Content.ReadFromJsonAsync<SpotifyUserResponse>()
                   ?? throw new HttpRequestException("Failed to deserialize user profile");

        var user = new SpotifyUser
        {
            Id = data.Id,
            DisplayName = data.DisplayName,
            Email = data.Email,
            ImageUrl = data.Images?.FirstOrDefault()?.Url,
        };

        return (user, activeToken);
    }

    // -------------------------------------------------------------------------
    // Secure token storage
    // -------------------------------------------------------------------------

    /// <inheritdoc />
    public async Task SaveTokenAsync(SpotifyToken token)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(token);
        await SecureStorage.SetAsync(SecureStoreKey, json);
    }

    /// <inheritdoc />
    public async Task<SpotifyToken?> GetStoredTokenAsync()
    {
        var raw = await SecureStorage.GetAsync(SecureStoreKey);
        if (string.IsNullOrEmpty(raw)) return null;
        return System.Text.Json.JsonSerializer.Deserialize<SpotifyToken>(raw);
    }

    /// <inheritdoc />
    public Task ClearTokenAsync()
    {
        SecureStorage.Remove(SecureStoreKey);
        return Task.CompletedTask;
    }

    // -------------------------------------------------------------------------
    // Internal JSON DTOs
    // -------------------------------------------------------------------------

    private sealed class SpotifyTokenResponse
    {
        [JsonPropertyName("access_token")]  public string AccessToken { get; set; } = string.Empty;
        [JsonPropertyName("refresh_token")] public string? RefreshToken { get; set; }
        [JsonPropertyName("expires_in")]    public int ExpiresIn { get; set; }
        [JsonPropertyName("scope")]         public string? Scope { get; set; }
    }

    private sealed class SpotifyUserResponse
    {
        [JsonPropertyName("id")]           public string Id { get; set; } = string.Empty;
        [JsonPropertyName("display_name")] public string DisplayName { get; set; } = string.Empty;
        [JsonPropertyName("email")]        public string Email { get; set; } = string.Empty;
        [JsonPropertyName("images")]       public SpotifyImage[]? Images { get; set; }
    }

    private sealed class SpotifyImage
    {
        [JsonPropertyName("url")] public string Url { get; set; } = string.Empty;
    }
}
